### ISSUE TITLE
False positive with 'unified-signatures'

### ISSUE THREAD

**Issue by stephanedr** [palantir/tslint#2016]
### Bug Report

- __TSLint version__: `4.3.1`
- __TypeScript version__: `2.1.4`
- __Running TSLint via__: VSCode

#### TypeScript code being linted

```ts
function test(): void;
function test(a: number, b: number): void;
function test(a?: number, b?: number): void {}
```

#### Actual behavior
These overloads can be combined into one signature with an optional parameter. (unified-signatures)'
at: '3,27'

#### Expected behavior
No warning. I want to allow either 0 or 2 parameters.

**Comment by nchen63** [comment]
Doesn't your 3rd signature allow 1 argument?

**Comment by nchen63** [comment]
@andy-ms

**Comment by stephanedr** [comment]
@nchen63 I considered that @andy-ms 's ticket title was answering your question.
In fact TS doesn't allow to call the implementation (`function test(a?: number, b?: number): void {}`), only the overload signatures can be called.

**Comment by davidwalschots** [comment]
I'm also failing to understand this rule, and I've therefore disabled it.

In the example below, I don't want to start merging overloads. For example, the second line has an optional country, but the third line doesn't, because it's required in stringified form by the `messageFn`. I don't want to merge these two as dictated by the rule, because that would allow not passing a country in that particular case.

```
export function postalCodeValidator(): ValidatorFn;
export function postalCodeValidator(message: string, country?: Country): ValidatorFn;
export function postalCodeValidator(messageFn: (country: string) => string, country: Country): ValidatorFn;
export function postalCodeValidator(message?: string | ((country: string) => string), country?: Country): ValidatorFn {

}
```

**Comment by andy-ms** [comment]
The failure there is referring to the first two overloads. Here's a simpler example:

```ts
export function f(): void;
export function f(a: string, b?: string): void;
export function f(...args: any[]): void {

}
```

The function can be passed 0, 1, or 2 parameters. So it could be just `f(a?: string, b?: string): void`.
